src/network/operator.rs

```diff
@@ -0,0 +1,335 @@
+                            if coinbase_records.len() == 0 {
+                                warn!("[ProposedBlock] Peer {} sent a candidate block with an incorrect owner.", peer_ip);
+                                return;
+                            }
+                            coinbase_records
+                        }
+                        Err(error) => {
+                            warn!("[PoolResponse] {}", error);
+                            return;
+                        }
+                    };
+
+                    // Ensure the block contains a difficulty that is at least the share difficulty.
+                    let proof_difficulty = {
+                        let proof_bytes = match block.header().proof() {
+                            Some(proof) => match proof.to_bytes_le() {
+                                Ok(bytes) => bytes,
+                                Err(error) => {
+                                    warn!("[PoolResponse] {}", error);
+                                    return;
+                                }
+                            },
+                            None => {
+                                warn!("[PoolResponse] Peer {} sent a candidate block with a missing proof.", peer_ip);
+                                return;
+                            }
+                        };
+
+                        let proof_difficulty = sha256d_to_u64(&proof_bytes);
+                        let share_difficulty = {
+                            let mut provers = self.provers.write().await;
+                            match provers.get(&prover_address) {
+                                Some((_, share_difficulty, _)) => *share_difficulty,
+                                None => {
+                                    provers.insert(prover_address, (Instant::now(), BASE_SHARE_DIFFICULTY, 0));
+                                    BASE_SHARE_DIFFICULTY
+                                }
+                            }
+                        };
+
+                        if proof_difficulty > share_difficulty {
+                            warn!("Block with insufficient share difficulty from {} ({})", peer_ip, prover_address);
+                            return;
+                        }
+
+                        proof_difficulty
+                    };
+
+                    // Update the score for the prover.
+                    // TODO: add round stuff
+                    // TODO: ensure shares can not be resubmitted
+                    if let Err(error) = self.state.add_shares(block.height(), &prover_address, 1) {
+                        error!("{}", error);
+                    }
+
+                    {
+                        // Update the internal state for this prover.
+                        let mut provers = self.provers.write().await;
+                        let mut prover = *provers.get_mut(&prover_address).expect("prover should have existing info");
```

Comment:
this should be an `if let Some(ref mut prover) = self.provers.write().await { ... }` block for greater readability; there can still be an `else` branch with a `panic`